pkg/snapshotcombiner: introduce SnapshotCombiner to handle combining and caching stats

flyth opened the pull request:
# Add SnapshotCombiner to handle combining and caching stats for multiple nodes

This PR adds a new and simple SnapshotCombiner library that takes stats of nodes and returns them combined on request. 
It adds a caching mechanism, so that when no stats get reported for a node and a certain amount of intervals, "old" stats will be used. This should increase user experience by not "flickering" if a node doesn't report its stats in time. The tradeoff in doing this, is that for one interval (or the ttl given in the settings), the results will not be 100% accurate (but they won't accurate as well if results are missing).

## How to use

You create a new SnapshotCombiner and add stats via `AddSnapshot()` for each node once the node reports stats. Periodically, you request the combined stats using `GetSnapshots()`.

Every time stats get added, a `ttl` (time-to-live) timer is reset for the reporting node and everytime stats are requested, all `ttl` values of all nodes will be decreased by one. Once a node reaches a ttl of zero, it won't be returned anymore.

Fixes #852

blanquicet commented on pkg/statscombiner/statscombiner_test.go:
```diff
@@ -0,0 +1,82 @@
+			IntervalStats:   map[string][]int{},
+			ExpectedEntries: 0,
+		},
+
```
```suggestion
```

flyth replied:
I think having the ExpectedEntries there despite having the default value helps reading the test cases, but I'm fine with removing it.

blanquicet replied:
I think it is not needed.

blanquicet commented on pkg/statscombiner/statscombiner_test.go:
```diff
@@ -0,0 +1,82 @@
+			IntervalStats:   map[string][]int{},
+			ExpectedEntries: 0,
+		},
+	}
```
Could you please add some tests with multiple keys?

blanquicet reviewed (commented):
I think the code looks excellent, you made it really simple and easy to follow! You found another great use case for generics 🚀 !

I am a little concerned about one case, but I am unsure if I am missing something. See my comments.

flyth commented on pkg/statscombiner/statscombiner.go:
```diff
@@ -0,0 +1,80 @@
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+
+	if curStats, ok := sc.stats[key]; ok {
+		curStats.stats = stats
+		curStats.ttl = sc.ttl
+		return
+	}
+	sc.stats[key] = &statsWrapper[T]{
+		stats: stats,
+		ttl:   sc.ttl,
+	}
```
You're right - those are the same. Initially I thought it would be a good idea to count the number of AddStats() calls as well, which required to differentiate the cases. I will remove that block. 👍

flyth commented on cmd/kubectl-gadget/top/top.go:
```diff
@@ -23,6 +23,8 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const DefaultStatsTTL = 2
```
The old stats should not be only shown when there are no new stats at all, but mixed with the new stats of other nodes. A concrete use case is the `top ebpf` where you measure the cumulative times of program runs. In case one node doesn't report in time, it would not magically make the not-updated entries disappear (removing the whole node from the output) but keep them for another interval.
Maybe we have to differentiate between the use cases...

blanquicet replied:
Mmmm ok, with the example of the `top ebpf` I got that you want to combine new and old stats on purpose so that the number of entries is always the same, not just to avoid the "empty screen" (no new stats at all). I am not sure, let me think about it.

@mauriciovasquezbernal @alban any comment?

alban replied:
I think it is fine like this for the normal "top" view.

But I am not sure about the "json" view. I wonder if we should give the raw output in the json view. Wdyt?

Also, I wonder about the benchmark use case where users run the benchmark for 10 seconds with `--timeout 10` and look at the final "cumul*" columns. If the last stats happen to be late, the benchmark will show cumul* values with 10% less. For this specific use case, we need to wait for the stats instead. Not sure how to fix that properly.

mauriciovasquezbernal replied:
I think there is not perfect solution for this issue. Maybe an alternative could be to introduce a sequence number on the data sent from each node and print all the stats from the same sequence number together? This would cause some delay on the client, as it would need to wait for all nodes, but maybe this works and avoids cases when old/new stats are reported together?

blanquicet commented on pkg/statscombiner/statscombiner.go:
```diff
@@ -0,0 +1,80 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statscombiner
```
I was wondering if it could have a more generic name. Maybe datacombiner or at least not related to "stats". In the end, it has nothing to do with stats. It is just combining `any` kind of data, don't you think?

flyth replied:
Hmm, that's true. It does however have this time component (ttl, expecting "events" and periodic requests). Naming it too generic might be confusing as the use case is pretty specific.

flyth replied:
Similarly to the answer below: what about `snapshotcombiner`, then? It simply overwrites old snapshots as new get in and keeps them only for a certain amount of (combined) requests.

blanquicet commented on pkg/statscombiner/statscombiner.go:
```diff
@@ -0,0 +1,80 @@
+type StatsCombiner[T any] struct {
+	lock  sync.Mutex
+	ttl   int
+	stats map[string]*statsWrapper[T]
```
I was wondering if we should rename the `stats` field. Maybe `statsBuffer`/`dataBuffer` or simply `buffer`? BTW, I think we should do the same [here](https://github.com/kinvolk/inspektor-gadget/blob/eea3fd22feaa689e1d2a23919bc27421a2200c21/pkg/gadgets/top/tcp/types/types.go#L87).

flyth replied:
What do you think about `snapshots`?

alban replied:
`snapshotsWithTTL`?

And a comment:
```
	// The keys are typically the node name
```

blanquicet commented on pkg/statscombiner/statscombiner_test.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	tests := []testRun{
+		{
```
Could you please add some additional "risky" tests? For instance, adding this one as first:

```golang
		{
			Name:            "try to get stats from a just initialised combiner",
			ExpectedEntries: 0,
		},
```

blanquicet commented on pkg/statscombiner/statscombiner_test.go:
```diff
@@ -0,0 +1,82 @@
+		},
+		{
+			Name:            "no node reports stats, old stats should still be there",
+			IntervalStats:   map[string][]int{},
```
```suggestion
```

This line is not needed. Same in similar tests.

blanquicet reviewed (commented):
Another quick round.

alban reviewed (commented):
Thanks!

alban commented:
The nodes use a ticker to get the stats every second:
```
IntervalDefault = 1
timer := time.NewTicker(t.config.Interval)
```

But according to the ticker documentation:
> The ticker will adjust the time interval or drop ticks to make up for slow receivers.

So it is possible that a node would completely skip a second. In addition, the stats might come late because of network latency, or because the gadget took more time to start on a specific node.

From the client perspective (the CLI), it is not possible to distinguish a late event from events that are skipped.

So I wonder if each `Event struct` should have a timestamp. The CLI can start the gadgets on each node and when it reaches a steady state on all nodes, it can start to gather results, grouping stats from the same second together, looking at the timestamp field (assuming each node's clocks are synchronised), and displaying the results with a couple of second delay (in case one node's stats come a bit late).

alban commented:
This PR improves the "top" UI a lot.

My concerns about benchmarks don't have to be addressed in this PR: I don't think they can be easily solved. Unless you can think of an easy way to solve them, I suggest to take this PR to fix the flickering problem, and the issues about benchmarking could be addressed later.

flyth commented:
> So I wonder if each Event struct should have a timestamp.

I like the idea of adding a timestamp - and if it'd be only for internal debugging purposes. We could also add a counter to keep track of the number of events sent.

Regarding the ticker dropping events (cited from the code):

> Give the channel a 1-element time buffer. If the client falls behind while reading, we drop ticks on the floor until the client catches up.

So while it surely can happen, I doubt that will be an issue, as the gadget itself would need to take longer than two intervals for its processing and/or the output channel's buffer would need to be filled (within that time).

> The CLI can start the gadgets on each node and when it reaches a steady state on all nodes, it can start to gather results, grouping stats from the same second together, looking at the timestamp field (assuming each node's clocks are synchronised), and displaying the results with a couple of second delay (in case one node's stats come a bit late).

I really don't think we should do a lot of synchronization; one failing node could lead to unexpected results. We should maybe indicate somehow that things are out of sync (like saying at the top "stats included from 4 out of 5 nodes (1 not responding, 1 outdated)" or even simpler), but I wouldn't go much father than that.

flyth commented:
I've renamed this to SnapshotCombiner, removed all references to "Stats" - and then, to confuse everyone, reintroduced "Stats" \o/. This time, however, the stats are being used to track how many snapshots are up to date, how many are known in total and how many have expired. These new stats are not displayed yet, though - I would like to leave that to another PR.
Also, I haven't updated/added the remaining tests, yet.

mauriciovasquezbernal commented on pkg/snapshotcombiner/statscombiner_test.go:
```diff
@@ -0,0 +1,76 @@
+		},
+		{
+			Name: "no node reports wrappedSnapshots, old wrappedSnapshots should have gotten deleted",
+		},
```
Maybe add some more steps with "node2" pushing updates?

mauriciovasquezbernal commented on pkg/snapshotcombiner/statscombiner_test.go:
```diff
@@ -0,0 +1,76 @@
+		ExpectedEntries int
+	}
+
+	tests := []testRun{
```
I was a bit confused at the beginning because I was thinking each of these was a different test, but those are "steps" of the same test that should run one after the other. Do you think there is a way to make it more clear?

mauriciovasquezbernal reviewed (commented):
Few minor comments around.

burak-ok reviewed (commented):
Thanks for simultaneously adding the unittest.
Always makes the usage and interface clearer :)

flyth commented on pkg/snapshotcombiner/snapshotcombiner.go:
```diff
@@ -0,0 +1,104 @@
+		entry.snapshot = snapshot
+		entry.ttl = sc.defaultTTL
+		entry.count++
+		entry.lastUpdate = time.Now()
```
Haha, locks, yeah... :) This isn't particularly hot code (at least for our use cases it's at the most called once per second per node), but yeah, I'll move the time.Now().

flyth commented on pkg/snapshotcombiner/snapshotcombiner.go:
```diff
@@ -0,0 +1,104 @@
+		Epochs: sc.epoch,
+	}
+
+	result := make([]T, 0)
```
Not particularly hot code as well (called once per interval, so at most once per second in total), but you're right, I'll change it.

mauriciovasquezbernal commented:
I think we can implement the combiner for fixing the "flickering" issue, but I'd say we need to figure out what to do with json. What do you think about the idea in https://github.com/inspektor-gadget/inspektor-gadget/pull/898#discussion_r961124875, is it too complex? 

Also, what about the case two updates come in the same interval? Is there a way to avoid overwriting them?

mauriciovasquezbernal reviewed (approved):
LGTM. Would love to see how it's used soon!
